onnx/defs/shape_inference.cc

```diff
@@ -249,16 +249,23 @@ void UnionTypeInfo(const TypeProto& source_type, TypeProto& target_type) {
           "Mismatched tensor element type:", " source=", source_elem_type, " target=", target_elem_type);
     }
 
-    UnionShapeInfoForTensor(source_type.tensor_type().shape(), *target_type.mutable_tensor_type());
+    if (source_type.tensor_type().has_shape()) {
+      UnionShapeInfoForTensor(source_type.tensor_type().shape(), *target_type.mutable_tensor_type());
+    } else {
+      target_type.mutable_tensor_type()->clear_shape();
+    }
   } else if (target_case == TypeProto::ValueCase::kSparseTensorType) {
     auto source_elem_type = source_type.sparse_tensor_type().elem_type();
     auto target_elem_type = target_type.sparse_tensor_type().elem_type();
     if (source_elem_type != target_elem_type) {
       fail_type_inference(
           "Mismatched sparse tensor element type:", " source=", source_elem_type, " target=", target_elem_type);
     }
-
-    UnionShapeInfoForTensor(source_type.sparse_tensor_type().shape(), *target_type.mutable_sparse_tensor_type());
+    if (source_type.sparse_tensor_type().has_shape()) {
+      UnionShapeInfoForTensor(source_type.sparse_tensor_type().shape(), *target_type.mutable_sparse_tensor_type());
+    } else {
+      target_type.mutable_sparse_tensor_type()->clear_shape();
+    }
```

Comment:
If UnionShapeInfoForTensor is not a public function, is it possible for us to change its interface to make it take source_type as an input instead of shape and then handle the has_shape() stuff inside the function?

Reply:
That sounds good. But, this function is exposed in the shape_inference.h file. We probably should not expose these functions there, but it is a bit too late ... we could deprecate it and remove it at some point, I suppose.

Reply:
I refactored the code as suggested, creating new overloads for `UnionShapeInfo`.